chapter-list: refactored and added new features

dyphire opened the pull request:
optimize scroll-list.lua:
- add format_header_string function for custom behaviour.
- add more methods and keybinds meant to control the list overlay.

refactor chapter-list.lua:
- add mouse control and options
- display the cursor position and the total number of lists in the header
- when the menu is first opened now the new behavior of list.selected is to select the current chapter
- reset cursor navigation when open the list, also add option to control whether this behavior is enabled(default)

dyphire commented:
@CogentRedTester 
Do you have any opinion on this PR?
I also created several new scroll menu scripts based on the new features. see [adevice-list.lua](https://github.com/dyphire/mpv-scripts/blob/main/adevice-list.lua), [track-menu.lua](https://github.com/dyphire/mpv-scripts/blob/main/track-menu.lua), [editions-notification-menu.lua](https://github.com/dyphire/mpv-scripts/blob/main/editions-notification-menu.lua)

dyphire commented:
In addition, I also found a problem that the `list.selected` will not be reset after switching files, resulting in its wrong state.

At present, I assign a new value to it to solve this problem. Do you have a better idea？

dyphire commented:
I have another question related to the above. When opening the scroll menu for the first time, the cursor stays on the currently used list item. At this time, move the cursor to other list items at will and close the menu.

After opening the menu again, the cursor at this time will inherit the state before closing. How to reset the cursor state to ensure that each opening stays on the currently used list item?

CogentRedTester reviewed (changes requested):
All the chapter-list changes are fine. I never planned to improve this script since it was just a demo of scroll-list, but I'm not opposed to some upgrades.

CogentRedTester commented:
> > How to reset the cursor state to ensure that each opening stays on the currently used list item?
> 
> Reference code from [jonniek/mpv-playlistmanager@b3b7582](https://github.com/jonniek/mpv-playlistmanager/commit/b3b75829bb77661370dddea929deccda89f59c71) , implement the cursor-pos reset function for chapter-list.lua.
> 
> As with reset function of `list.selected` after switching files, neither seems can be implemented in scroll-list.lua.

Yes those aren't and shouldn't be added to `scroll-list`. We have no way of knowing what scroll-list will be used for, it's up to the scripts using scroll-list to implement those features as you have done in chapter-list.

CogentRedTester commented on examples/chapter-list.lua:
```diff
@@ -20,32 +41,58 @@ local function open_chapter()
     for i = 1, #chapter_list do
         local item = {}
-        if (i-1 == curr_chapter) then
+        if (i - 1 == curr_chapter) then
+            list.selected = curr_chapter + 1
             item.style = [[{\c&H33ff66&}]]
         end
```
```suggestion
```

Instead of burying this in the loop we can place it somewhere more prominent at the end.

CogentRedTester commented on examples/chapter-list.lua:
```diff
@@ -20,32 +41,58 @@ local function open_chapter()
-        item.ass = item.ass..'\\h\\h\\h'..list.ass_escape(chapter_list[i].title)
+        item.ass = item.ass..'\\h\\h\\h'..list.ass_escape(title)
         list.list[i] = item
     end
```
```suggestion
    end
    
    list.selected = curr_chapter + 1
    list.list[curr_chapter + 1].style = [[{\c&H33ff66&}]]
```

dyphire replied:
There is a problem with the change here, the `curr_chapter` may be a nil value. It is better to keep it unchanged.

CogentRedTester replied:
Ah you're right, we can put it in an if block then: `if curr_chapter then`

dyphire replied:
> Ah you're right, we can put it in an if block then: `if curr_chapter then`

This is fine, but I want to stay in the loop as a better example script. 
Because in some usage scenarios,  the `list.selected` must be set in the loop. You can refer to another script of mine, such as [adevice-list.lua](https://github.com/dyphire/mpv-scripts/blob/main/adevice-list.lua).

dyphire commented on examples/chapter-list.lua:
```diff
@@ -20,32 +41,58 @@ local function open_chapter()
+
+mp.register_script_message("toggle-chapter-browser", function()
+    list:toggle()
+    reset_cursor()
```
it better

dyphire commented on scroll-list.lua:
```diff
@@ -43,6 +43,16 @@ function scroll_list.ass_escape(str, replace_newline)
+		:gsub("%%total%%", #self.list)
+    else string = string:gsub("%[.*%]", "") end
+    return string
+end
```
You are right. The scroll-list.lua should be generic and put the function implementation into the chapter-list.lua.

CogentRedTester commented:
Also change the commit message to `chapter-list: refactored and added new features` and include at least a basic list of the new features added (key options, new selector behaviour, etc)

CogentRedTester commented on README.md:
```diff
@@ -75,12 +75,14 @@ Changing these can break the script if certain function calls are missing. Make
 | Method                   | Description                                                                                                                    |
 |--------------------------|--------------------------------------------------------------------------------------------------------------------------------|
-| format_header()          | formats and prints the header to the overlay                                                                                   |
+| format_header_string()   | format and return the header                                                                                                  |
```
```suggestion
| format_header_string()   | format and return the header string - allows one to modify or substitute the header string on each redraw |
```

dyphire commented:
> Also change the commit message to and include at least a basic list of the new features added (key options, new selector behaviour, etc)`chapter-list: refactored and added new features`

Attempted to rewrite the commit message

dyphire commented:
@CogentRedTester Do you have any views on the current changes?

CogentRedTester commented on scroll-list.lua:
```diff
@@ -43,6 +43,11 @@ function scroll_list.ass_escape(str, replace_newline)
 end
 
+--format and return the header string
+function scroll_list:format_header_string(string)
+    return string
```
```suggestion
function scroll_list:format_header_string(str)
    return str
```

CogentRedTester commented on examples/chapter-list.lua:
```diff
@@ -1,51 +1,117 @@
-    end
+--escape header specifies the format
+--display the cursor position and the total number of lists in the header
+function list:format_header_string(string)
```
```suggestion
function list:format_header_string(str)
```
Don't use `string` as a variable, it interferes with the default string library. It's fine here but it introduces confusion.

CogentRedTester reviewed (changes requested):
Hey sorry, I forgot to re-review this PR. I just have a couple of very minor changes, but overall it's looking good.

dyphire commented on examples/chapter-list.lua:
```diff
@@ -1,51 +1,117 @@
+function list:format_header_string(string)
+    if #list.list > 0 then
+        string = string:gsub("%%cursor%%", list.selected)
+		:gsub("%%total%%", #list.list)
```
It works.

CogentRedTester commented:
Looks good, I'll merge when you're ready.

dyphire commented:
> Looks good, I'll merge when you're ready.

If you have no comments on the commit message, then I am ready.
